ops/charm.py

```diff
@@ -261,62 +344,157 @@ class RelationCreatedEvent(RelationEvent):
+    settings. They should not error if the settings are incomplete,
+    since it can be guaranteed that when the remote unit changes its
+    settings, the event will be fire again.
+
+    The settings that be queried, or set, are determined by the
+    relation’s interface.
     """
 
 
 class RelationDepartedEvent(RelationEvent):
-    """Represents the `relation-departed` hook from Juju.
-
-    This is the inverse of the RelationJoinedEvent, representing when a unit
-    is leaving the relation (the unit is being removed, the app is being removed,
-    the relation is being removed). It is fired once for each unit that is
-    going away.
+    """Event triggered when a unit leaves a relation.
+
+    This is the inverse of the :class: `RelationJoinedEvent`,
+    representing when a unit is leaving the relation (the unit is
+    being removed, the app is being removed, the relation is being
+    removed). It is fired once for each unit that is going away.
+
+    When the remote unit is known to be leaving the relation, this
+    will result in the :class: `RelationChangedEvent` firing at least
+    once, after which the :class: `RelationDepartedEvent` will
+    fire. The :class: `RelationDepartedEvent` will fire once
+    only. Once the :class: `RelationDepartedEvent` has fired no
+    further :class: `RelationChangedEvent` events will fire.
+
+    Callback methods bound to this event may be used to remove all
+    references to the departing remote unit, because there’s no
+    guarantee that it’s still part of the system; it’s perfectly
+    probable (although not guaranteed) that the system running that
+    unit has already shut down.
+
+    Once all callback methods bound to this event have been run for
+    such a relation, the unit agent will fire, the :class:
+    `RelationBrokenEvent` event.
     """
 
 
 class RelationBrokenEvent(RelationEvent):
-    """Represents the `relation-broken` hook from Juju.
-
-    If a relation is being removed (`juju remove-relation` or `juju remove-application`),
-    once all the units have been removed, RelationBrokenEvent will fire to signal
-    that the relationship has been fully terminated.
+    """Event triggered when a relation is removed.
+
+    If a relation is being removed (`juju remove-relation` or `juju
+    remove-application`), once all the units have been removed, this
+    event will fire to signal that the relationship has been fully
+    terminated.
+
+    The event indicates that the current relation is no longer valid,
+    and that the charm’s software must be configured as though the
+    relation had never existed. It will only be called after every
+    callback method bound to :class: `RelationDepartedEvent` event has
+    been run. If a callback method bound to this event is being
+    executed, it is gauranteed that no remote units are currently
+    known locally.
     """
 
 
 class StorageEvent(HookEvent):
```

Comment:
This is as far as I have gotten in this review.

Reply:
Awesome. Take a break. Have a beer. I am getting hungry too, and will call it a day soon but may hang out on IRC much later. I really appreciate you taking the time to do this.